### PR TITLE
Better implementation of PSR-4 Namespace rules

### PR DESCRIPTION
**Pull request by aleferri**
Consider src to be the base directory (PSR-4 definition) and the "surplus" leading segments as parts of the namespace prefix.
Implementation is rough. For it to be better, i would need to add a configuration options in the project to let the user set the namespace prefix; fetch the configuration in this Checker and remove the appropriate leading part of the namespace. Unfortunately i don't know how to do that.

### CODE REVIEW

**Comment by junichi11** [comment]
Can you submit it to the JIRA and add an example project there?

**Comment by aleferri** [comment]
Did create the issue as improvements
https://issues.apache.org/jira/browse/NETBEANS-6180

**Comment by junichi11** [comment]
@aleferri I'm sorry I'm late... Maybe, I overlooked the notification.

`src` is hardcoded. So, I think we should get the source directory.
Unit tests are missing.

File: php/php.editor/src/org/netbeans/modules/php/editor/verification/PSR4Hint.java [hunk 1]
```diff
@@ -102,6 +102,9 @@ public void visit(NamespaceDeclaration node) {
                         if (CancelSupport.getDefault().isCancelled()) {
                             return;
                         }
+                        if (currentDir.getName().equals("src")) { // PSR-4 require corrispondence to the directory structure only for the contiguos subnamespaces
```
Comment (tmysik) [inline php/php.editor/src/org/netbeans/modules/php/editor/verification/PSR4Hint.java:105]: Typos in the comment.

**Comment by tmysik** [comment]
Agreed, some unit test(s) would be fine. Thanks.

**Comment by aleferri** [comment]
@junichi11 How do i get the environment inside the Hint? My original idea was to search for the top folder, but i couldn't figure how to get objects outside of the current file.

**Comment by junichi11** [comment]
I would the the source directory like the following:

```java
PhpModule phpModule = PhpModule.Factory.forFileObject(fileObject);
if (phpModule != null) {
    FileObject sourceDirectory = phpModule.getSourceDirectory();
}
```

**Comment by junichi11** [comment]
BTW, you should create a PR with not the master branch but the new branch.

**Comment by tmysik** [comment]
@junichi11 Good catch.

@aleferri Please, create a branch in your clone, containing the number of the ticket and ideally also a few words, from the ticket title, typically (e.g. `1234-fix-npe`).

Thanks.

**Comment by aleferri** [comment]
Ok, i am going to open a new PR then

**Comment by aleferri** [comment]
Sorry to ask, but how i can run the tests? -Dcluster.config=php build netbeans, but the tests didn't run.

**Comment by matthiasblaesing** [comment]
Either from the UI Kontextmenu of the project or on CLI: `ant -f php/php.editor/build.xml test`.

**Comment by aleferri** [comment]
Tests goes in timeout

**Comment by matthiasblaesing** [comment]
@aleferri which tests timeout? Your local run? Which test.

With flaky tests I regularly run the tests before my change, note which fail and run after. I consider my changes "sane" if the same set of tests succeeds and fails (and yes in perfect world this would not be necessary, but...)

**Comment by neilcsmith-net** [comment]
@junichi11 @tmysik marked stale and removed from NB16 milestone for now. Please follow up as you see fit - re-add, re-review or close.

**Comment by tmysik** [comment]
It seems to me that (all?) the comments are still valid.

**Comment by aleferri** [comment]
The implementation doesn't work, as it simply disable the warning even when the thing is wrong. I did that test by building the whole thing and opening my test project and also a should-fail one (aka i did a manual test).
Unfortunately i am out of ideas on how to fix it and automated tests still don't run (not even one of them, all of them go in timeout).

**Comment by neilcsmith-net** [comment]
I hadn't noticed how small the actual proposed changes were either - busy going through closing or marking stale PRs with milestones that hadn't been completed.  Closing for discussion elsewhere then.

**Comment by junichi11** [comment]
Please submit it as a new PR with a new branch. Thanks!

**Pull request closed without merging**